File: ictyb_back/src/main/java/kepco/prorject/ictyb/back/ictyb_back/work_opinion/controller/WorkOpinionController.java
package kepco.prorject.ictyb.back.ictyb_back.work_opinion.controller;

import com.fasterxml.jackson.databind.ObjectMapper;

import kepco.prorject.ictyb.back.ictyb_back.common.BaseResponse;
import kepco.prorject.ictyb.back.ictyb_back.common.enums.ResultCodeEnum;
import kepco.prorject.ictyb.back.ictyb_back.common.enums.StatusEnum;
import kepco.prorject.ictyb.back.ictyb_back.work_opinion.model.WorkOpinionDto;
import kepco.prorject.ictyb.back.ictyb_back.work_opinion.service.WorkOpinionService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;
import org.springframework.web.multipart.MultipartHttpServletRequest;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.List;

@Slf4j
@RestController
@RequestMapping("/api/work_opinion")
@RequiredArgsConstructor
public class WorkOpinionController {

    private final WorkOpinionService workOpinionService;
    private final ObjectMapper objectMapper;

    /**
     * 업무 협의 목록 조회 (지시번호 기준)
     * GET /api/work_opinion/v1.0/list/{instrNo}
     */
    @GetMapping("/v1.0/list/{instrNo}")
    public ResponseEntity<BaseResponse<?>> getDiscussions(@PathVariable String instrNo) {
        try {
            List<WorkOpinionDto.DiscussionItem> data = workOpinionService.getDiscussions(instrNo);
            return ResponseEntity.ok(BaseResponse.builder()
                    .status(StatusEnum.SUCCESS)
                    .data(data)
                    .resultCode(ResultCodeEnum.SUCCESS)
                    .resultMsg("협의 목록을 조회했습니다.")
                    .build());
        } catch (Exception e) {
            log.error("협의 목록 조회 오류 instrNo={}", instrNo, e);
            return ResponseEntity.ok(BaseResponse.builder()
                    .status(StatusEnum.FAIL)
                    .resultCode(ResultCodeEnum.INTERNAL_SERVER_ERROR)
                    .resultMsg("협의 목록 조회 중 오류가 발생했습니다.")
                    .build());
        }
    }

    /**
     * 새 협의 스레드 생성
     * POST /api/work_opinion/v1.0 (multipart/form-data)
     * - discussionData: WorkOpinionDto.CreateDiscussionReq (JSON)
     * - files: 첨부파일 (선택, 0개 이상)
     */
    @PostMapping(value = "/v1.0", consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
    public ResponseEntity<BaseResponse<?>> createDiscussion(
            @RequestPart("discussionData") String discussionDataJson,
            MultipartHttpServletRequest multipartRequest) {
        try {
            WorkOpinionDto.CreateDiscussionReq req =
                    objectMapper.readValue(discussionDataJson, WorkOpinionDto.CreateDiscussionReq.class);
            List<MultipartFile> files = multipartRequest.getMultiFileMap().get("files");
            WorkOpinionDto.DiscussionItem data = workOpinionService.createDiscussion(req, files);
            return ResponseEntity.ok(BaseResponse.builder()
                    .status(StatusEnum.SUCCESS)
                    .data(data)
                    .resultCode(ResultCodeEnum.SUCCESS_CREATE)
                    .resultMsg("협의가 등록됐습니다.")
                    .build());
        } catch (Exception e) {
            log.error("협의 생성 오류", e);
            return ResponseEntity.ok(BaseResponse.builder()
                    .status(StatusEnum.FAIL)
                    .resultCode(ResultCodeEnum.INTERNAL_SERVER_ERROR)
                    .resultMsg("협의 등록 중 오류가 발생했습니다.")
                    .build());
        }
    }

    /**
     * 댓글 등록
     * POST /api/work_opinion/v1.0/comment (multipart/form-data)
     * - commentData: WorkOpinionDto.CreateCommentReq (JSON)
     * - files: 첨부파일 (선택, 0개 이상)
     */
    @PostMapping(value = "/v1.0/comment", consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
    public ResponseEntity<BaseResponse<?>> addComment(
            @RequestPart("commentData") String commentDataJson,
            MultipartHttpServletRequest multipartRequest) {
        try {
            WorkOpinionDto.CreateCommentReq req =
                    objectMapper.readValue(commentDataJson, WorkOpinionDto.CreateCommentReq.class);
            List<MultipartFile> files = multipartRequest.getMultiFileMap().get("files");
            WorkOpinionDto.CommentItem data = workOpinionService.addComment(req, files);
            return ResponseEntity.ok(BaseResponse.builder()
                    .status(StatusEnum.SUCCESS)
                    .data(data)
                    .resultCode(ResultCodeEnum.SUCCESS_CREATE)
                    .resultMsg("댓글이 등록됐습니다.")
                    .build());
        } catch (Exception e) {
            log.error("댓글 등록 오류", e);
            return ResponseEntity.ok(BaseResponse.builder()
                    .status(StatusEnum.FAIL)
                    .resultCode(ResultCodeEnum.INTERNAL_SERVER_ERROR)
                    .resultMsg("댓글 등록 중 오류가 발생했습니다.")
                    .build());
        }
    }

    /**
     * 댓글 첨부파일 다운로드
     * GET /api/work_opinion/v1.0/attach/download?cmntId=&seqNo=
     */
    @GetMapping("/v1.0/attach/download")
    public ResponseEntity<?> downloadAttach(
            @RequestParam String cmntId,
            @RequestParam Long seqNo) {
        try {
            WorkOpinionDto.DownloadFile file = workOpinionService.downloadAttach(cmntId, seqNo);
            String encoded = URLEncoder.encode(
                    file.getRealFileName() != null ? file.getRealFileName() : "file",
                    StandardCharsets.UTF_8
            ).replace("+", "%20");

            return ResponseEntity.ok()
                    .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename*=UTF-8''" + encoded)
                    .contentType(MediaType.APPLICATION_OCTET_STREAM)
                    .body(file.getResource());
        } catch (Exception e) {
            log.error("댓글 첨부파일 다운로드 오류 cmntId={} seqNo={}", cmntId, seqNo, e);
            return ResponseEntity.notFound().build();
        }
    }
}

File: ictyb_front/app/hooks/work_opinion/WorkOpinionController.ts
import type { BaseResponse } from '@hooks/common/base';
import type {
  DiscussionItem,
  CommentItem,
  CreateDiscussionReq,
  CreateCommentReq,
} from './type';
import apiClient from '@hooks/common/service/clientService';

/** 지시번호 기준 협의 목록 + 댓글 조회 */
export const fetchDiscussions = async (instrNo: string): Promise<DiscussionItem[]> => {
  const res = await apiClient.get<BaseResponse<DiscussionItem[]>>(
    `/api/work_opinion/v1.0/list/${instrNo}`,
  );
  return res.data.data ?? [];
};

/** 새 협의 스레드 생성 (첨부파일 포함) */
export const createDiscussion = async (
  req: CreateDiscussionReq,
  files: File[] = [],
): Promise<DiscussionItem> => {
  const formData = new FormData();
  formData.append('discussionData', new Blob([JSON.stringify(req)], { type: 'application/json' }));
  files.forEach((f) => formData.append('files', f));

  const res = await apiClient.post<BaseResponse<DiscussionItem>>(
    '/api/work_opinion/v1.0',
    formData,
    { headers: { 'Content-Type': 'multipart/form-data' } },
  );
  return res.data.data;
};

/** 댓글 등록 (첨부파일 포함) */
export const addComment = async (
  req: CreateCommentReq,
  files: File[] = [],
): Promise<CommentItem> => {
  const formData = new FormData();
  formData.append('commentData', new Blob([JSON.stringify(req)], { type: 'application/json' }));
  files.forEach((f) => formData.append('files', f));

  const res = await apiClient.post<BaseResponse<CommentItem>>(
    '/api/work_opinion/v1.0/comment',
    formData,
    { headers: { 'Content-Type': 'multipart/form-data' } },
  );
  return res.data.data;
};

/** 지시번호에 연결된 모든 협의 스레드를 로그인 사용자 기준으로 일괄 읽음 처리 */
export const markDiscussionsRead = async (instrNo: string): Promise<void> => {
  await apiClient.post<BaseResponse<void>>(`/api/work_opinion/v1.0/${instrNo}/read`);
};

/** 댓글 첨부파일을 브라우저를 통해 즉시 다운로드합니다. */
export const downloadCommentAttach = (cmntId: string, seqNo: number, realFileName: string) => {
  const url = `http://localhost:8082/api/work_opinion/v1.0/attach/download?cmntId=${cmntId}&seqNo=${seqNo}`;
  const link = document.createElement('a');
  link.href = url;
  link.download = realFileName;
  link.click();
};
